### PR TITLE
Elasticsearch mappings cannot be changed

### LINKED ISSUE #24363: Elasticsearch mappings cannot be changed
### Summary (*)
Magento 2.3.2
As a developer, I need to develop a custom module that affects the search functionality. This requires extra data to be indexed in elasticsearch. This module requires an extra dynamic template sent to the elasticsearch server in the mappings.  

### Examples (*)
in version 2.3.2, the mappings are sent via `Magento\Elasticsearch6\Model\Client\Elasticsearch::addFieldsMapping` (or any other models that can be used as elasticsearch clients depending on elasticsearch version)
And the dynamic templates are hardocded

```
'dynamic_templates' => [
                        [
                            'price_mapping' => [
                                'match' => 'price_*',
                                'match_mapping_type' => 'string',
                                'mapping' => [
                                    'type' => 'float',
                                    'store' => true,
                                ],
                            ],
                        ],
                        ....
```
The value cannot be changed via plugins, because the value is sent to elasticsearch as soon as it it built. The method that actually sends the mappings `$this->getClient()->indices()->putMapping($params);` cannot be pluginized because the object that is being called is not created via Object manager.  

### Proposed solution

A method that just returns an empty array in the elasticsearch client. The result of this method is merged with the mappings sent to elasticsearch before sending them. This method can be pluginized by third party modules.  
This will not break backwards compatibility.

### PR DESCRIPTION
**Pull request by engcom-Kilo**
### Fixed Issues (if relevant)

1. Fixes https://github.com/magento/magento2/issues/24363

### Manual testing scenarios (*)

1. see https://github.com/magento/magento2/issues/24363

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by engcom-Kilo** [comment]
@magento run all tests

File: app/code/Magento/Elasticsearch7/etc/di.xml [hunk 1]
```diff
@@ -221,6 +221,47 @@
                 <item name="elasticsearch7_copy_searchable_fields_to_search_field" xsi:type="object">Magento\Elasticsearch\Model\Adapter\FieldMapper\CopySearchableFieldsToSearchField</item>
                 <item name="elasticsearch7_add_default_search_field" xsi:type="object">Magento\Elasticsearch\Model\Adapter\FieldMapper\AddDefaultSearchField</item>
             </argument>
+            <argument name="dynamicTemplates" xsi:type="array">
```
Comment (zakdma) [inline app/code/Magento/Elasticsearch7/etc/di.xml]: I would recommend create mapping params processors pull (Composite pattern) and pass params into it.
So any developer can add own processor that can change parameters as it wish.
Also there is problem with such structures you created that we cannot replace or delete items from it, only add new.

**Review by zakdma — CHANGES_REQUESTED** [review]

**Comment by engcom-Kilo** [comment]
@magento run all tests

File: app/code/Magento/Elasticsearch7/etc/di.xml (hunk 1 above)
Comment (engcom-Kilo) [reply app/code/Magento/Elasticsearch7/etc/di.xml]: Done

**Comment by engcom-Kilo** [comment]
@magento run all tests

File: app/code/Magento/Elasticsearch7/Test/Unit/Model/Client/ElasticsearchTest.php [hunk 2]
```diff
@@ -93,12 +98,21 @@ protected function setUp(): void
             ->willReturn(['version' => ['number' => '7.0.0']]);
 
         $this->objectManager = new ObjectManagerHelper($this);
+        $dynamicTemplatesProvider = new DynamicTemplatesProvider(
```
Comment (zakdma) [inline app/code/Magento/Elasticsearch7/Test/Unit/Model/Client/ElasticsearchTest.php:101]: You should get it using ObjectManger. It is more safe in terms of extensibility and also no need to provide constructor paramaters in this case

**Review by zakdma — CHANGES_REQUESTED** [review]

File: app/code/Magento/Elasticsearch7/Test/Unit/Model/Client/ElasticsearchTest.php (hunk 2 above)
Comment (engcom-Kilo) [reply app/code/Magento/Elasticsearch7/Test/Unit/Model/Client/ElasticsearchTest.php:101]: Object Manager for unit test will create dynamic template provider with mock mappers with processTemplates() method return null.
Also, Object Manager already deprecated
https://github.com/magento/magento2/blob/2.4-develop/lib/internal/Magento/Framework/TestFramework/Unit/Helper/ObjectManager.php#L16

**Review by zakdma — APPROVED** [review]

**Comment by engcom-Kilo** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run Functional Tests CE, Functional Tests B2B

**Comment by zakdma** [comment]
@magento import code to https://github.com/magento-tsg/magento2ce

**Comment by magento-engcom-team** [comment]
@zakdma the branch with code successfully imported into`magento-tsg/magento2ce` repository. Branch name: `imported-magento-magento2-32755`.

**Pull request merged**